--- src/Terraria/Terraria/Gore.cs
+++ src/TerraAngel/Terraria/Gore.cs
@@ -272,21 +_,6 @@
     private float ChumFloatingChunk_GetWaterLine(int X, int Y)
     {
         float result = position.Y + Height;
-        if (Main.tile[X, Y - 1] == null)
-        {
-            Main.tile[X, Y - 1] = new Tile();
-        }
-
-        if (Main.tile[X, Y] == null)
-        {
-            Main.tile[X, Y] = new Tile();
-        }
-
-        if (Main.tile[X, Y + 1] == null)
-        {
-            Main.tile[X, Y + 1] = new Tile();
-        }
-
         if (Main.tile[X, Y - 1].liquid > 0)
         {
             result = Y * 16;
@@ -661,16 +_,6 @@
                 int num12 = (int)(vector.Y / 16f);
                 _ = position.Y / 16f;
                 int num13 = (int)((position.Y + Height) / 16f);
-                if (Main.tile[num11, num12] == null)
-                {
-                    Main.tile[num11, num12] = new Tile();
-                }
-
-                if (Main.tile[num11, num13] == null)
-                {
-                    Main.tile[num11, num13] = new Tile();
-                }
-
                 if (velocity.Y > 0f)
                 {
                     velocity.Y *= 0.5f;
@@ -1413,6 +_,7 @@
 
     public static int NewGore(Vector2 Position, Vector2 Velocity, int Type, float Scale = 1f)
     {
+        if (OptimizationTool.OptimizationToolCache?.DisableDust ?? false) return 600;
         if (Main.netMode == 2)
         {
             return 600;
