--- src/Terraria/Terraria/Item.cs
+++ src/TerraAngel/Terraria/Item.cs
@@ -1090,7 +_,9 @@
         return false;
     }
 
+    public string AffixName() => TerraAngel.Hooks.DrawHooks.AffixNameHook((x) => x._AffixName(), this);
+    
-    public string AffixName()
+    public string _AffixName()
     {
         if (prefix <= 0 || prefix >= Lang.prefix.Length)
         {
@@ -48786,7 +_,7 @@
         }
 
         RebuildTooltip();
-        if (type > 0 && type < ItemID.Count && ItemID.Sets.Deprecated[type])
+        if (type > 0 && type < ItemID.Count && ItemID.Sets.Deprecated[type] && !ClientConfig.Settings.IsPreserveDeprecatedItems)
         {
             TurnToAir();
         }
