--- src/Terraria/Terraria/MessageBuffer.cs
+++ src/TerraAngel/Terraria/MessageBuffer.cs
@@ -3,6 +_,7 @@
 using System.IO;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics.PackedVector;
+using TerraAngel;
 using Terraria.Audio;
 using Terraria.Chat;
 using Terraria.DataStructures;
@@ -233,7 +_,8 @@
                     Main.myPlayer = num91;
                     Player player9 = Main.player[num91];
                     NetMessage.TrySendData(4, -1, -1, null, num91);
+                    if (ClientConfig.Settings.IsSendClientUUID)
-                    NetMessage.TrySendData(68, -1, -1, null, num91);
+                        NetMessage.TrySendData(68, -1, -1, null, num91);
                     NetMessage.TrySendData(16, -1, -1, null, num91);
                     NetMessage.TrySendData(42, -1, -1, null, num91);
                     NetMessage.TrySendData(50, -1, -1, null, num91);
@@ -2690,42 +_,9 @@
                     int num15 = reader.ReadInt16();
                     byte b3 = reader.ReadByte();
                     byte liquidType = reader.ReadByte();
-                    if (Main.netMode == 2 && Netplay.SpamCheck)
-                    {
-                        int num16 = whoAmI;
-                        int num17 = (int)(Main.player[num16].position.X + (float)(Main.player[num16].width / 2));
-                        int num18 = (int)(Main.player[num16].position.Y + (float)(Main.player[num16].height / 2));
-                        int num19 = 10;
-                        int num20 = num17 - num19;
-                        int num21 = num17 + num19;
-                        int num22 = num18 - num19;
-                        int num23 = num18 + num19;
-                        if (num14 < num20 || num14 > num21 || num15 < num22 || num15 > num23)
-                        {
-                            Netplay.Clients[whoAmI].SpamWater += 1f;
-                        }
-                    }
-
-                    if (Main.tile[num14, num15] == null)
-                    {
-                        Main.tile[num14, num15] = new Tile();
-                    }
-
-                    lock (Main.tile[num14, num15])
-                    {
                         Main.tile[num14, num15].liquid = b3;
                         Main.tile[num14, num15].liquidType(liquidType);
-                        if (Main.netMode == 2)
-                        {
-                            WorldGen.SquareTileFrame(num14, num15);
-                            if (b3 == 0)
-                            {
-                                NetMessage.SendData(48, -1, whoAmI, null, num14, num15);
-                            }
-                        }
-
                         break;
-                    }
                 }
 
             case 49:
@@ -4086,6 +_,8 @@
                         Vector2 vector = Vector2.Normalize(value3);
                         if (!vector.HasNaNs())
                         {
+                            if (!ClientConfig.Settings.DisableNebulaLagPacket)
+                            {
                             int type12 = 90;
                             float num128 = 0f;
                             float num129 = (float)Math.PI / 15f;
@@ -4116,6 +_,7 @@
                                 num130 = (Main.dust[num133].fadeIn = num130 + num131);
                                 Main.dust[num133].velocity += vector * 1.5f;
                             }
+                            }
                         }
 
                         player11.NebulaLevelup(num126);
@@ -4324,6 +_,11 @@
                         if (Main.netMode == 2)
                         {
                             NetMessage.SendPlayerHurt(num24, playerDeathReason2, damage, num25, flag, pvp2, num26, -1, whoAmI);
+                        }
+                        
+                        if (damage == 1 && playerDeathReason2._sourceNPCIndex == 203)
+                        {
+                            Main.player[num26].TerraAngelUser = true;
                         }
                     }
 
