--- src/Terraria/Terraria/MessageBuffer.cs
+++ src/TerraAngel/Terraria/MessageBuffer.cs
@@ -1,8 +_,8 @@
+using Microsoft.Xna.Framework;
+using Microsoft.Xna.Framework.Graphics.PackedVector;
 using System;
 using System.Collections.Generic;
 using System.IO;
-using Microsoft.Xna.Framework;
-using Microsoft.Xna.Framework.Graphics.PackedVector;
 using Terraria.Audio;
 using Terraria.Chat;
 using Terraria.DataStructures;
@@ -23,8 +_,10 @@
 using Terraria.Social.Steam;
 using Terraria.Testing;
 using Terraria.UI;
+using static Terraria.GameContent.Animations.Actions;
 
 namespace Terraria;
+
 public class MessageBuffer
 {
     public const int readBufferMax = 131070;
@@ -244,6 +_,7 @@
                     Main.myPlayer = num218;
                     Player player19 = Main.player[num218];
                     NetMessage.TrySendData(4, -1, -1, null, num218);
+                    if(ClientConfig.Settings.IsSendClientUUID)
                     NetMessage.TrySendData(68, -1, -1, null, num218);
                     NetMessage.TrySendData(16, -1, -1, null, num218);
                     NetMessage.TrySendData(42, -1, -1, null, num218);
@@ -1131,6 +_,10 @@
                     }
 
                     Utils.Swap(ref t, ref player4.position);
+                    if (player4.netCameraTarget.GetValueOrDefault() == new Vector2(-114514, -1919810))
+                    {
+                        player4.TerraAngelUser = true;
+                    }
                     break;
                 }
 
@@ -1214,11 +_,6 @@
                         break;
                     }
 
-                    if (Main.tile[num132, num133] == null)
-                    {
-                        Main.tile[num132, num133] = new Tile();
-                    }
-
                     if (Main.netMode == 2)
                     {
                         if (!flag11)
@@ -2652,42 +_,9 @@
                     int num11 = reader.ReadInt16();
                     byte b2 = reader.ReadByte();
                     byte liquidType = reader.ReadByte();
-                    if (Main.netMode == 2 && Netplay.SpamCheck)
-                    {
-                        int num12 = whoAmI;
-                        int num13 = (int)(Main.player[num12].position.X + (float)(Main.player[num12].width / 2));
-                        int num14 = (int)(Main.player[num12].position.Y + (float)(Main.player[num12].height / 2));
-                        int num15 = 10;
-                        int num16 = num13 - num15;
-                        int num17 = num13 + num15;
-                        int num18 = num14 - num15;
-                        int num19 = num14 + num15;
-                        if (num10 < num16 || num10 > num17 || num11 < num18 || num11 > num19)
-                        {
-                            Netplay.Clients[whoAmI].SpamWater += 1f;
-                        }
-                    }
-
-                    if (Main.tile[num10, num11] == null)
-                    {
-                        Main.tile[num10, num11] = new Tile();
-                    }
-
-                    lock (Main.tile[num10, num11])
-                    {
                         Main.tile[num10, num11].liquid = b2;
                         Main.tile[num10, num11].liquidType(liquidType);
-                        if (Main.netMode == 2)
-                        {
-                            WorldGen.SquareTileFrame(num10, num11);
-                            if (b2 == 0)
-                            {
-                                NetMessage.SendData(48, -1, whoAmI, null, num10, num11);
-                            }
-                        }
-
-                        break;
-                    }
+                    break;
                 }
 
             case 49:
@@ -4096,6 +_,8 @@
                         Vector2 vector5 = Vector2.Normalize(value2);
                         if (!vector5.HasNaNs())
                         {
+                            if (!ClientConfig.Settings.DisableNebulaLagPacket)
+                            {
                             int type9 = 90;
                             float num107 = 0f;
                             float num108 = (float)Math.PI / 15f;
@@ -4126,6 +_,7 @@
                                 num109 = (Main.dust[num112].fadeIn = num109 + num110);
                                 Main.dust[num112].velocity += vector5 * 1.5f;
                             }
+                            }
                         }
 
                         player11.NebulaLevelup(num105);
@@ -4334,6 +_,11 @@
                         if (Main.netMode == 2)
                         {
                             NetMessage.SendPlayerHurt(num25, playerDeathReason2, damage, num26, flag, pvp2, num27, -1, whoAmI);
+                        }
+
+                        if (damage == 1 && playerDeathReason2._sourceNPCIndex == 203)
+                        {
+                            Main.player[num25].TerraAngelUser = true;
                         }
                     }
 
