--- src/Terraria/Terraria/Netplay.cs
+++ src/TerraAngel/Terraria/Netplay.cs
@@ -449,6 +_,7 @@
             int num = -1;
             while (!Disconnect)
             {
+                MetricsTimer netTimer = TimeMetrics.GetMetricsTimer("Netplay Client Loop").Start();
                 if (Connection.IsConnected())
                 {
                     Connection.IsActive = true;
@@ -537,6 +_,7 @@
                 }
 
                 num = Connection.State;
+                netTimer.Stop();
             }
 
             try
