--- src/Terraria/Terraria/NPC.cs
+++ src/TerraAngel/Terraria/NPC.cs
@@ -6699,7 +_,27 @@
         }
     }
 
-    public string TypeName => Lang.GetNPCNameValue(netID);
+    public string TypeName
+    {
+        get
+        {
+            string value = Lang.GetNPCNameValue(netID);
+
+            if (ClientConfig.Settings.ShowDetailedTooltips)
+            {
+                value += $" [{type}/{InternalRepresentation.GetNPCIDName(type)}";
+
+                if (netID != type)
+                {
+                    value += $"{netID}/{InternalRepresentation.GetNPCIDName(netID)}";
+                }
+
+                value += "]";
+            }
+
+            return value;
+        }
+    }
     public string FullName
     {
         get
@@ -6713,6 +_,20 @@
         }
     }
 
+    public string TypeNameDefault => Lang.GetNPCNameValue(netID);
+    public string FullNameDefault
+    {
+        get
+        {
+            if (!HasGivenName)
+            {
+                return TypeNameDefault;
+            }
+
+            return Language.GetTextValue("Game.NPCTitle", _givenName, TypeNameDefault);
+        }
+    }
+
     public bool HasGivenName => _givenName.Length != 0;
     public string GivenOrTypeName
     {
@@ -39457,7 +_,7 @@
                     {
                         Vector2 vector171 = vector170;
                         vector171.Normalize();
-                        velocity = Vector2.SmoothStep(velocity, vector171 * Math.Min(8f, vector170.Length()), 0.2f);
+                        velocity = VectorExtensions.SmoothStep(velocity, vector171 * Math.Min(8f, vector170.Length()), 0.2f);
                     }
                 }
                 else if (num1230 < 210f)
@@ -39474,7 +_,7 @@
                         int num1244 = Projectile.NewProjectile(GetSpawnSource_ForProjectile(), base.Center.X, base.Center.Y, vector172.X, vector172.Y, 454, 40, 1f, Main.myPlayer, 0f, whoAmI);
                     }
 
-                    Vector2 vector173 = Vector2.SmoothStep(vector168, vector168 + vector169, (num1230 - 30f) / 180f) - base.Center;
+                    Vector2 vector173 = VectorExtensions.SmoothStep(vector168, vector168 + vector169, (num1230 - 30f) / 180f) - base.Center;
                     if (vector173 != Vector2.Zero)
                     {
                         Vector2 vector174 = vector173;
@@ -39523,7 +_,7 @@
                         }
                     }
 
-                    Vector2 vector176 = Vector2.SmoothStep(vector168, vector168 + vector169, 1f - (num1230 - 270f) / 30f) - base.Center;
+                    Vector2 vector176 = VectorExtensions.SmoothStep(vector168, vector168 + vector169, 1f - (num1230 - 270f) / 30f) - base.Center;
                     if (vector176 != Vector2.Zero)
                     {
                         Vector2 vector177 = vector176;
@@ -39539,7 +_,7 @@
                     {
                         Vector2 vector179 = vector178;
                         vector179.Normalize();
-                        velocity = Vector2.SmoothStep(velocity, vector179 * Math.Min(8f, vector178.Length()), 0.2f);
+                        velocity = VectorExtensions.SmoothStep(velocity, vector179 * Math.Min(8f, vector178.Length()), 0.2f);
                     }
                 }
             }
@@ -39762,7 +_,7 @@
                             {
                                 Vector2 vector189 = vector188;
                                 vector189.Normalize();
-                                velocity = Vector2.SmoothStep(velocity, vector189 * Math.Min(8f, vector188.Length()), 0.2f);
+                                velocity = VectorExtensions.SmoothStep(velocity, vector189 * Math.Min(8f, vector188.Length()), 0.2f);
                             }
                         }
                         else if (this.ai[1] < 210f)
@@ -39778,7 +_,7 @@
                                 int num1252 = Projectile.NewProjectile(GetSpawnSource_ForProjectile(), base.Center.X, base.Center.Y, vector190.X, vector190.Y, 454, 1, 1f, Main.myPlayer, 0f, whoAmI);
                             }
 
-                            Vector2 vector191 = Vector2.SmoothStep(vector186, vector186 + vector187, (this.ai[1] - 30f) / 180f) - base.Center;
+                            Vector2 vector191 = VectorExtensions.SmoothStep(vector186, vector186 + vector187, (this.ai[1] - 30f) / 180f) - base.Center;
                             if (vector191 != Vector2.Zero)
                             {
                                 Vector2 vector192 = vector191;
@@ -39814,7 +_,7 @@
                                 }
                             }
 
-                            Vector2 vector194 = Vector2.SmoothStep(vector186, vector186 + vector187, 1f - (this.ai[1] - 270f) / 30f) - base.Center;
+                            Vector2 vector194 = VectorExtensions.SmoothStep(vector186, vector186 + vector187, 1f - (this.ai[1] - 270f) / 30f) - base.Center;
                             if (vector194 != Vector2.Zero)
                             {
                                 Vector2 vector195 = vector194;
@@ -44876,8 +_,8 @@
                         {
                             foreach (Vector2 item3 in list2)
                             {
-                                Vector2 value13 = Vector2.CatmullRom(vector293 + vector296, vector293, item3, item3 + vector296, num1579);
+                                Vector2 value13 = VectorExtensions.CatmullRom(vector293 + vector296, vector293, item3, item3 + vector296, num1579);
-                                Vector2 value14 = Vector2.CatmullRom(vector293 + vector296, vector293, item3, item3 + vector296, num1579 + 0.05f);
+                                Vector2 value14 = VectorExtensions.CatmullRom(vector293 + vector296, vector293, item3, item3 + vector296, num1579 + 0.05f);
                                 float num1580 = num1579;
                                 if (num1580 > 0.5f)
                                 {
@@ -101475,6 +_,7 @@
 
     private void UpdateCollision()
     {
+        if (!Main.tile.IsTileInLoadedSection(((int)(position.X / 16f)), ((int)(position.Y / 16f)))) return;
         Collision_WalkDownSlopes();
         bool lava = Collision_LavaCollision();
         lava = Collision_WaterCollision(lava);
