--- src/Terraria/Terraria/Player.cs
+++ src/TerraAngel/Terraria/Player.cs
@@ -37,6 +_,7 @@
 using Terraria.UI.Gamepad;
 using Terraria.Utilities;
 using Terraria.WorldBuilding;
+using TerraAngel.Tools;
 
 namespace Terraria;
 public class Player : Entity, IFixLoadedData
@@ -718,6 +_,8 @@
     public bool active;
     public bool host;
     public Minecart.Customization MinecartSettings;
+
+    public bool TerraAngelUser = false;
     public int emoteTime;
     public CreativeUnlocksTracker creativeTracker;
     private static byte[] ENCRYPTION_KEY = new UnicodeEncoding().GetBytes("h3y_gUyZ");
@@ -1110,8 +_,10 @@
     public int potionDelay;
     public byte difficulty;
     public byte wetSlime;
-    public HitTile hitTile;
+    public HitTile? _hitTile;
-    public HitTile hitReplace;
+    public HitTile? _hitReplace;
+    public HitTile hitTile { get { if (_hitTile is null) _hitTile = new HitTile(); return _hitTile; } }
+    public HitTile hitReplace { get { if (_hitReplace is null) _hitReplace = new HitTile(); return _hitReplace; } }
     public int jump;
     public int head = -1;
     public int body = -1;
@@ -11407,6 +_,7 @@
 
     public void Yoraiz0rEye()
     {
+        if ((OptimizationTool.OptimizationToolCache?.DisableDust ?? false) || ClientConfig.Settings.IgnoreReLogicBullshit) return;
         int num = 0;
         num += bodyFrame.Y / 56;
         if (num >= Main.OffsetsPlayerHeadgear.Length)
@@ -13895,7 +_,7 @@
             shinyStone = true;
         }
 
-        if (currentItem.type == 4989)
+        if (currentItem.type == 4989 || ToolManager.GetTool<InfiniteFlightTool>().Enabled)
         {
             empressBrooch = true;
             moveSpeed += 0.075f;
@@ -17685,7 +_,7 @@
                 }
 
                 int num5 = tile.frameY / 22;
-                if (num5 < TorchID.Count)
+                if (num5 < TorchID.Count && num5 >= 0)
                 {
                     nearbyTorch[num5] = true;
                     if (num5 == 17 && (tile.liquid == 0 || tile.liquidType() != 0))
@@ -18232,7 +_,9 @@
         return (int)((float)damage * (1f + num3));
     }
 
+    public void ResetEffects() => TerraAngel.Hooks.PlayerHooks.PlayerResetEffectsHook((x) => x._ResetEffects(), this);
+
-    public void ResetEffects()
+    private void _ResetEffects()
     {
         IsAllowedToHoldItems = true;
         if (rapidAttackBonus > 0f)
@@ -24141,7 +_,7 @@
 
     public static int GetFloorTileType(int x, int y)
     {
-        return ((int?)GetFloorTile(x, y)?.type) ?? (-1);
+        return (int)GetFloorTile(x, y).type;
     }
 
     private void MakeFloorDust(bool Falling, int type, int frameX, int frameY, int paintColor)
@@ -28904,9 +_,12 @@
             CheckCrackedBrickBreak();
         }
 
+        if ((!ToolManager.GetTool<NoClipTool>().Enabled && i == Main.myPlayer) || (i != Main.myPlayer && Main.tile.IsTileInLoadedSection(((int)(position.X / 16f)), ((int)(position.Y / 16f)))))
+        {
         if (!shimmering)
         {
             SlopeDownMovement();
+            // patch context
             bool flag27 = mount.Type == 7 || mount.Type == 8 || mount.Type == 12 || mount.Type == 44 || mount.Type == 49;
             bool flag28 = mount.Type == 48;
             if (!mount.Active)
@@ -28919,20 +_,24 @@
             {
                 Collision.StepDown(ref position, ref velocity, width, height, ref stepSpeed, ref gfxOffY, (int)gravDir, waterWalk || waterWalk2);
             }
-
+            // patch context
             if (gravDir == -1f)
             {
+                // patch context
                 if ((carpetFrame != -1 || velocity.Y <= gravity) && !controlUp)
                 {
                     Collision.StepUp(ref position, ref velocity, width, height, ref stepSpeed, ref gfxOffY, (int)gravDir, controlUp);
                 }
+                // patch context
             }
             else if ((carpetFrame != -1 || velocity.Y >= gravity) && !controlDown && !IsRidingTracks && !flag27 && grappling[0] == -1)
             {
                 Collision.StepUp(ref position, ref velocity, width, height, ref stepSpeed, ref gfxOffY, (int)gravDir, controlUp);
             }
+            // patch context
         }
 
+        // patch context
         oldPosition = position;
         oldDirection = direction;
         bool falling = false;
@@ -29157,20 +_,22 @@
         {
             FloorVisuals(falling);
         }
-
+        // patch context
         if (whoAmI == Main.myPlayer && !shimmering)
         {
             Collision.SwitchTiles(position, width, height, oldPosition, 5);
         }
-
+        // patch context
         PressurePlateHelper.UpdatePlayerPosition(this);
         BordersMovement();
         numMinions = 0;
         slotsMinions = 0f;
+        // patch context
         if (Main.netMode != 2 && mount.Type != 8)
         {
             ItemCheck_ManageRightClickFeatures();
         }
+        }
 
         ItemCheckWrapped(i);
         PlayerFrame();
@@ -32800,7 +_,9 @@
         }
     }
 
+    private void ItemCheckWrapped(int i) => TerraAngel.Hooks.PlayerHooks.PlayerItemCheckHook((x, y) => x._ItemCheckWrapped(y), this, i);
+
-    private void ItemCheckWrapped(int i)
+    private void _ItemCheckWrapped(int i)
     {
         int num;
         if (!controlLeft && !controlRight && !controlUp && !controlDown && PlayerInput.UsingGamepad && Main.SmartCursorIsUsed)
@@ -34450,14 +_,16 @@
         if (Recipe.TileUsedInRecipes[type])
         {
             int baseOption = 0;
-            MapHelper.GetTileBaseOption(myX, myY, type, Main.tile[myX, myY], ref baseOption);
+            ref TileData tile = ref Main.tile.GetTileRef(myX, myY);
+            MapHelper.GetTileBaseOption(myX, myY, type, ref tile, ref baseOption);
             return new NewCraftingUI.CraftStationRecipeFilter(type, baseOption);
         }
 
         if (TileID.Sets.CountsAsWaterForCrafting[type])
         {
             int baseOption2 = 0;
-            MapHelper.GetTileBaseOption(myX, myY, type, Main.tile[myX, myY], ref baseOption2);
+            ref TileData tile = ref Main.tile.GetTileRef(myX, myY);
+            MapHelper.GetTileBaseOption(myX, myY, type, ref tile, ref baseOption2);
             return new NewCraftingUI.WaterSourceRecipeFilter(type, baseOption2);
         }
 
@@ -38975,6 +_,17 @@
 
     public void Teleport(Vector2 newPos, int Style = 0, int extraInfo = 0)
     {
+        
+        if (ToolManager.GetTool<TeleportTool>().LogTargetPosition)
+        {
+            ClientLoader.Console.WriteLine(GetString($"Teleported to {newPos.ToString()} (Style: {Style}, ExtraInfo: {extraInfo})"));
+        }
+        
+        if (!ToolManager.GetTool<TeleportTool>().AllowTeleport)
+        {
+            return;
+        }
+        
         try
         {
             _funkytownAchievementCheckCooldown = 100;
@@ -39289,6 +_,8 @@
             Hooks.EnterWorld(whoAmI);
         }
 
+        TerraAngel.Hooks.PlayerHooks.presenceUpdateCount = 0;
+
         if (whoAmI == Main.myPlayer)
         {
             _localMinionRespawner.RestoreMinionsFor(this);
@@ -39631,6 +_,17 @@
 
     public double Hurt(PlayerDeathReason damageSource, int Damage, int hitDirection, bool pvp = false, bool quiet = false, bool Crit = false, int cooldownCounter = -1, bool dodgeable = true)
     {
+        if (whoAmI == Main.myPlayer && ToolManager.GetTool<AntiHurtTool>().Enabled)
+        {
+            statLife = statLifeMax2;
+            if (ToolManager.GetTool<AntiHurtTool>().FramesSinceLastLifePacket == 0)
+            {
+                ToolManager.GetTool<AntiHurtTool>().FramesSinceLastLifePacket = 6;
+                NetMessage.SendData(MessageID.PlayerLife, -1, -1, null, whoAmI);
+            }
+            return 0.0d;
+        }
+
         if (shimmering && AllowShimmerDodge(damageSource, cooldownCounter, dodgeable))
         {
             return 0.0;
@@ -40306,6 +_,12 @@
 
     public void KillMe(PlayerDeathReason damageSource, double dmg, int hitDirection, bool pvp = false)
     {
+        if (whoAmI == Main.myPlayer && ToolManager.GetTool<AntiHurtTool>().Enabled)
+        {
+            NetMessage.SendData(MessageID.PlayerLife, -1, -1, null, whoAmI);
+            return;
+        }
+
         if (creativeGodMode || (DebugOptions.PracticeMode && DebugUtils.PracticeModeReset(this, damageSource)) || dead)
         {
             return;
@@ -41153,13 +_,23 @@
 
     private void PlaceThing_Walls()
     {
-        if (inventory[selectedItem].createWall < 0 || !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange))
+        bool blockedByRangeCheck = !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange);
+
+        if (ToolManager.GetTool<InfiniteReachTool>().Enabled || BuilderModeTool.Enabled)
+            blockedByRangeCheck = false;
+
+        if (inventory[selectedItem].createWall < 0 || blockedByRangeCheck)
         {
             return;
         }
 
+        bool wallPlaceBlockedByNoFreeSpace = (!Main.tile[tileTargetX + 1, tileTargetY].active() && Main.tile[tileTargetX + 1, tileTargetY].wall <= 0 && !Main.tile[tileTargetX - 1, tileTargetY].active() && Main.tile[tileTargetX - 1, tileTargetY].wall <= 0 && !Main.tile[tileTargetX, tileTargetY + 1].active() && Main.tile[tileTargetX, tileTargetY + 1].wall <= 0 && !Main.tile[tileTargetX, tileTargetY - 1].active() && Main.tile[tileTargetX, tileTargetY - 1].wall <= 0);
+
+        if (BuilderModeTool.Enabled)
+            wallPlaceBlockedByNoFreeSpace = false;
+
         cursorItemIconEnabled = true;
-        if (!ItemTimeIsZero || itemAnimation <= 0 || !controlUseItem || (!Main.tile[tileTargetX + 1, tileTargetY].active() && Main.tile[tileTargetX + 1, tileTargetY].wall <= 0 && !Main.tile[tileTargetX - 1, tileTargetY].active() && Main.tile[tileTargetX - 1, tileTargetY].wall <= 0 && !Main.tile[tileTargetX, tileTargetY + 1].active() && Main.tile[tileTargetX, tileTargetY + 1].wall <= 0 && !Main.tile[tileTargetX, tileTargetY - 1].active() && Main.tile[tileTargetX, tileTargetY - 1].wall <= 0) || Main.tile[tileTargetX, tileTargetY].wall == inventory[selectedItem].createWall)
+        if (((!ItemTimeIsZero || itemAnimation <= 0) && !BuilderModeTool.Enabled) || !controlUseItem || wallPlaceBlockedByNoFreeSpace || Main.tile[tileTargetX, tileTargetY].wall == inventory[selectedItem].createWall)
         {
             return;
         }
@@ -41291,7 +_,13 @@
     {
         Item item = inventory[selectedItem];
         int tileToCreate = item.createTile;
-        if (tileToCreate < 0 || !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, item.tileBoost + blockRange) || !SmartCursorHelper.TileTargetDesired())
+
+        bool blockedByRangeCheck = !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, item.tileBoost + blockRange);
+
+        if (ToolManager.GetTool<InfiniteReachTool>().Enabled || BuilderModeTool.Enabled)
+            blockedByRangeCheck = false;
+
+        if (tileToCreate < 0 || blockedByRangeCheck || !SmartCursorHelper.TileTargetDesired())
         {
             return;
         }
@@ -41321,7 +_,7 @@
             }
         }
 
-        if (canUse && ((!tile.active() && !PlaceThing_Tiles_IsBlockedByLava(item.createTile, item.placeStyle, tile)) || PlaceThing_IsReplacableBlock(tile) || tileToCreate == 199 || tileToCreate == 23 || tileToCreate == 662 || tileToCreate == 661 || tileToCreate == 2 || tileToCreate == 109 || tileToCreate == 60 || tileToCreate == 70 || tileToCreate == 633 || Main.tileMoss[tileToCreate]) && ItemTimeIsZero && itemAnimation > 0 && controlUseItem)
+        if (canUse && ((!tile.active() && !PlaceThing_Tiles_IsBlockedByLava(item.createTile, item.placeStyle, tile)) || PlaceThing_IsReplacableBlock(tile) || tileToCreate == 199 || tileToCreate == 23 || tileToCreate == 662 || tileToCreate == 661 || tileToCreate == 2 || tileToCreate == 109 || tileToCreate == 60 || tileToCreate == 70 || tileToCreate == 633 || Main.tileMoss[tileToCreate]) && (ItemTimeIsZero && itemAnimation > 0 || BuilderModeTool.Enabled) && controlUseItem)
         {
             bool? overrideCanPlace = null;
             int? forcedRandom = null;
@@ -41364,6 +_,9 @@
             canPlace = PlaceThing_Tiles_BlockPlacementForAssortedThings(canPlace);
         }
 
+        if (BuilderModeTool.Enabled)
+            canPlace = true;
+
         if (canPlace)
         {
             PlaceThing_Tiles_PlaceIt(newObjectType, data, tileToCreate);
@@ -43555,7 +_,12 @@
 
     private void PlaceThing_PaintScrapper()
     {
-        if (ItemID.Sets.IsPaintScraper[inventory[selectedItem].type] && IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange))
+        bool allowedByRangeCheck = IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange);
+
+        if (ToolManager.GetTool<InfiniteReachTool>().Enabled || BuilderModeTool.Enabled)
+            allowedByRangeCheck = true;
+
+        if (ItemID.Sets.IsPaintScraper[inventory[selectedItem].type] && allowedByRangeCheck)
         {
             int num = tileTargetX;
             int num2 = tileTargetY;
@@ -43575,7 +_,7 @@
         }
 
         cursorItemIconEnabled = true;
-        if (!ItemTimeIsZero || itemAnimation <= 0 || !controlUseItem)
+        if (((!ItemTimeIsZero || itemAnimation <= 0) && !BuilderModeTool.Enabled) || !controlUseItem)
         {
             return;
         }
@@ -43630,7 +_,7 @@
         }
 
         cursorItemIconEnabled = true;
-        if (ItemTimeIsZero && itemAnimation > 0 && controlUseItem)
+        if ((BuilderModeTool.Enabled || ItemTimeIsZero && itemAnimation > 0) && controlUseItem)
         {
             if (WorldGen.paintTile(x, y, 0, broadCast: true) || WorldGen.paintCoatTile(x, y, 0, broadcast: true))
             {
@@ -43645,7 +_,12 @@
 
     private void PlaceThing_PaintRoller()
     {
-        if ((inventory[selectedItem].type != 1072 && inventory[selectedItem].type != 1544) || !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange))
+        bool blockedByRangeCheck = !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange);
+
+        if (ToolManager.GetTool<InfiniteReachTool>().Enabled || BuilderModeTool.Enabled)
+            blockedByRangeCheck = false;
+
+        if ((inventory[selectedItem].type != 1072 && inventory[selectedItem].type != 1544) || blockedByRangeCheck)
         {
             return;
         }
@@ -43655,7 +_,7 @@
         if (Main.tile[num, num2] != null && Main.tile[num, num2].wall > 0)
         {
             cursorItemIconEnabled = true;
-            if (ItemTimeIsZero && itemAnimation > 0 && controlUseItem)
+            if ((BuilderModeTool.Enabled || ItemTimeIsZero && itemAnimation > 0) && controlUseItem)
             {
                 TryPainting(num, num2, paintingAWall: true);
             }
@@ -43664,7 +_,12 @@
 
     private void PlaceThing_Paintbrush()
     {
-        if ((inventory[selectedItem].type != 1071 && inventory[selectedItem].type != 1543) || !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange))
+        bool blockedByRangeCheck = !IsInTileInteractionRange(tileTargetX, tileTargetY, TileReachCheckSettings.Simple, inventory[selectedItem].tileBoost + blockRange);
+
+        if (ToolManager.GetTool<InfiniteReachTool>().Enabled || BuilderModeTool.Enabled)
+            blockedByRangeCheck = false;
+
+        if ((inventory[selectedItem].type != 1071 && inventory[selectedItem].type != 1543) || blockedByRangeCheck)
         {
             return;
         }
@@ -43674,7 +_,7 @@
         if (Main.tile[num, num2] != null && Main.tile[num, num2].active())
         {
             cursorItemIconEnabled = true;
-            if (ItemTimeIsZero && itemAnimation > 0 && controlUseItem)
+            if ((BuilderModeTool.Enabled || ItemTimeIsZero && itemAnimation > 0) && controlUseItem)
             {
                 TryPainting(num, num2);
             }
@@ -44051,7 +_,7 @@
         return num;
     }
 
-    private void Fishing_GetBait(out int baitPower, out int baitType)
+    public void Fishing_GetBait(out int baitPower, out int baitType)
     {
         baitPower = 0;
         baitType = 0;
@@ -45111,6 +_,7 @@
                 lastVisualizedSelectedItem = item.Clone();
             }
 
+            bool noConsumeInBuilderMode = BuilderModeTool.Enabled && (item.createTile >= 0 || item.createWall >= 0);
             if (whoAmI == Main.myPlayer)
             {
                 if (!dontConsumeWand && itemTimeMax != 0 && itemTime == itemTimeMax && item.tileWand > 0)
@@ -45175,7 +_,7 @@
 
                     if (flag7)
                     {
-                        if (item.stack > 0)
+                        if (item.stack > 0 && !noConsumeInBuilderMode)
                         {
                             item.stack--;
                         }
@@ -45188,7 +_,7 @@
                     }
                 }
 
-                if (item.stack <= 0 && itemAnimation == 0)
+                if (item.stack <= 0 && itemAnimation == 0 && !noConsumeInBuilderMode)
                 {
                     inventory[selectedItem] = new Item();
                     itemAnimationMax = 0;
@@ -59004,8 +_,6 @@
         usedGummyWorm = false;
         usedAmbrosia = false;
         Array.Clear(adjTile, 0, adjTile.Length);
-        hitTile = new HitTile();
-        hitReplace = new HitTile();
         mount = new Mount();
         talkNPC = -1;
         piggyBankProjTracker.Clear();
