--- src/Terraria/Terraria/Program.cs
+++ src/TerraAngel/Terraria/Program.cs
@@ -8,6 +_,7 @@
 using System.Threading;
 using ReLogic.IO;
 using ReLogic.OS;
+using TerraAngel;
 using Terraria.Initializers;
 using Terraria.Localization;
 using Terraria.Social;
@@ -16,8 +_,8 @@
 namespace Terraria;
 public static class Program
 {
-    public static bool IsXna = true;
+    public static bool IsXna = false;
-    public static bool IsFna = false;
+    public static bool IsFna = true;
     public static bool IsMono = Type.GetType("Mono.Runtime") != null;
     public const bool IsDebug = false;
     public static Dictionary<string, string> LaunchParameters = new Dictionary<string, string>();
@@ -56,7 +_,7 @@
 
     public static void ForceLoadThread(object threadContext)
     {
-        ForceLoadAssembly(Assembly.GetExecutingAssembly(), initializeStaticMembers: true);
+        ForceLoadAssembly(typeof(Main).Assembly, initializeStaticMembers: true);
         LoadedEverything = true;
     }
 
@@ -186,6 +_,7 @@
 
         LaunchParameters = Utils.ParseArguements(args);
         SavePath = (LaunchParameters.ContainsKey("-savedirectory") ? LaunchParameters["-savedirectory"] : Platform.Get<IPathService>().GetStoragePath("Terraria"));
+        ClientLoader.LoadClient();
         ThreadPool.SetMinThreads(8, 8);
         InitializeConsoleOutput();
         SetupLogging();
@@ -219,10 +_,15 @@
         }
 
         using Main main = new Main();
+#if !DEBUG
         try
         {
+#endif
             Lang.InitializeLegacyLocalization();
+            if (ClientConfig.Settings.UseSteamSocialAPI)
-            SocialAPI.Initialize();
+                SocialAPI.Initialize();
+            else
+                SocialAPI.Initialize(SocialMode.None);
             LaunchInitializer.LoadParameters(main);
             Main.OnEnginePreload += StartForceLoad;
             if (Main.dedServ)
@@ -231,14 +_,16 @@
             }
 
             main.Run();
+#if !DEBUG
         }
         catch (Exception e)
         {
             DisplayException(e);
         }
+#endif
     }
 
-    private static void TrySettingDefaultFNADriver(string[] args)
+    public static void TrySettingDefaultFNADriver(string[] args)
     {
         bool flag = false;
         for (int i = 0; i < args.Length; i++)
@@ -256,7 +_,7 @@
         }
     }
 
-    private static void DisplayException(Exception e)
+    public static void DisplayException(Exception e)
     {
         try
         {
@@ -284,7 +_,7 @@
                 Console.WriteLine(Language.GetTextValue("Error.ServerCrash"), DateTime.Now, text);
             }
 
-            MessageBox.Show(text, "Terraria: Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
+            SDL3.SDL.SDL_ShowSimpleMessageBox(SDL3.SDL.SDL_MessageBoxFlags.SDL_MESSAGEBOX_ERROR, "Terraria: Error", text, Main.instance.Window.Handle);
         }
         catch
         {
