--- src/Terraria/Terraria/Projectile.cs
+++ src/TerraAngel/Terraria/Projectile.cs
@@ -38878,8 +_,8 @@
         Vector2 value = master.MountedCenter + new Vector2(master.direction * 10, 0f);
         Vector2 tangent = new Vector2(0f, -100f);
         Vector2 tangent2 = new Vector2(0f, -100f);
-        Vector2 vector = Vector2.Hermite(value, tangent, base.Center, tangent2, num2);
+        Vector2 vector = VectorExtensions.Hermite(value, tangent, base.Center, tangent2, num2);
-        Vector2 vector2 = Vector2.Hermite(value, tangent, base.Center, tangent2, num3);
+        Vector2 vector2 = VectorExtensions.Hermite(value, tangent, base.Center, tangent2, num3);
         for (float num4 = 0f; num4 <= 1f; num4 += 0.1f)
         {
             Dust dust = Dust.NewDustPerfect(Vector2.Lerp(vector, vector2, num4), 306, (vector2 - vector) * 0.1f, 0, new Color(100, 227, 255, 127), 1f + num3);
@@ -45693,7 +_,7 @@
             }
 
             float amount = MathHelper.Lerp(value, value2, Utils.GetLerpValue(num, 30f, timeLeft, clamped: true));
-            velocity = Vector2.SmoothStep(velocity, value3, amount);
+            velocity = VectorExtensions.SmoothStep(velocity, value3, amount);
             if (type == 931 || type == 1039)
             {
                 velocity *= MathHelper.Lerp(0.85f, 1f, Utils.GetLerpValue(0f, 90f, timeLeft, clamped: true));
@@ -46814,7 +_,7 @@
         Vector2 vector2 = -velocity;
         Vector2 value = center + vector2 * 2f;
         Vector2 value2 = vector + vector2 * (1f - num2 * 3f);
-        base.Center = Vector2.CatmullRom(value, center, vector, value2, num2);
+        base.Center = VectorExtensions.CatmullRom(value, center, vector, value2, num2);
         int num3 = type;
         if (num3 == 856)
         {
@@ -46984,7 +_,7 @@
         }
 
         ai[0] += 1f / ai[1];
-        Vector2 vector3 = Vector2.CatmullRom(amount: Utils.Clamp(ai[0], 0f, 1f), value1: value, value2: vector, value3: vector2, value4: value2);
+        Vector2 vector3 = VectorExtensions.CatmullRom(amount: Utils.Clamp(ai[0], 0f, 1f), value1: value, value2: vector, value3: vector2, value4: value2);
         if (flag)
         {
             for (float num3 = 0f; num3 < 1f; num3 += 0.5f)
@@ -52055,7 +_,7 @@
                 AI_GetMyGroupIndexAndFillBlackList(blacklist, out var index2, out var totalIndexesInGroup2);
                 AI_156_GetIdlePosition(index2, totalIndexesInGroup2, out var idleSpot2, out var _);
                 velocity = Vector2.Zero;
-                base.Center = Vector2.SmoothStep(base.Center, idleSpot2, 0.45f);
+                base.Center = VectorExtensions.SmoothStep(base.Center, idleSpot2, 0.45f);
                 if (Main.rand.Next(20) == 0)
                 {
                     int num6 = AI_156_TryAttackingNPCs(blacklist);
@@ -52078,7 +_,7 @@
             AI_GetMyGroupIndexAndFillBlackList(blacklist, out var index3, out var totalIndexesInGroup3);
             AI_156_GetIdlePosition(index3, totalIndexesInGroup3, out var idleSpot3, out var idleRotation3);
             velocity = Vector2.Zero;
-            base.Center = Vector2.SmoothStep(base.Center, idleSpot3, 0.45f);
+            base.Center = VectorExtensions.SmoothStep(base.Center, idleSpot3, 0.45f);
             rotation = rotation.AngleLerp(idleRotation3, 0.45f);
             if (Main.rand.Next(20) == 0)
             {
@@ -79440,7 +_,7 @@
                                 Vector2 value13 = vector73.RotatedBy(((num904 == 0) ? 1f : (-1f)) * ((float)Math.PI * 2f) / (num899 * 2f));
                                 for (float num905 = 0f; num905 < num902; num905++)
                                 {
-                                    Vector2 vector74 = Vector2.SmoothStep(vector73, value13, num905 / num902);
+                                    Vector2 vector74 = VectorExtensions.SmoothStep(vector73, value13, num905 / num902);
                                     float num906 = MathHelper.SmoothStep(num900, num901, num905 / num902);
                                     int num907 = Dust.NewDust(new Vector2(position.X, position.Y), 6, 6, 134, 0f, 0f, 100, default(Color), 1.3f);
                                     Dust dust2 = Main.dust[num907];
