--- src/Terraria/Terraria/Utils.cs
+++ src/TerraAngel/Terraria/Utils.cs
@@ -539,11 +_,6 @@
                 continue;
             }
 
-            if (Main.tile[num4, num5] == null)
-            {
-                Main.tile[num4, num5] = new Tile();
-            }
-
             Tile tile = Main.tile[num4, num5];
             if ((settings.avoidWalls && tile.wall > 0) || (tile.wall == 87 && !NPC.downedPlantBoss) || (Main.wallDungeon[tile.wall] && (double)num5 > Main.worldSurface && !NPC.downedBoss3) || !CheckForGoodTeleportationSpot_CheckNoInvalidTiles(num4, num5, settings))
             {
@@ -560,11 +_,6 @@
                     break;
                 }
 
-                if (Main.tile[num4, num5 + num8] == null)
-                {
-                    Main.tile[num4, num5 + num8] = new Tile();
-                }
-
                 Tile tile2 = Main.tile[num4, num5 + num8];
                 vector = new Vector2(num4, num5 + num8) * 16f + new Vector2(-num6 / 2 + 8, -num2);
                 Collision.SlopeCollision(vector, teleporteeVelocity, num6, num2, teleporteeGravityDirection);
@@ -1163,7 +_,8 @@
 
     public static Vector2 ScreenToWorldPosition(this Vector2 screenPosition)
     {
-        return Vector2.Transform(screenPosition * Main.UIScale, Matrix.Invert(Main.GameViewMatrix.TransformationMatrix)) + Main.screenPosition;
+        Matrix.Invert(Main.GameViewMatrix.TransformationMatrix, out var invertedTransformMatrix);
+        return Vector2.Transform(screenPosition * Main.UIScale, invertedTransformMatrix) + Main.screenPosition;
     }
 
     public static string PrettifyPercentDisplay(float percent, string originalFormat)
@@ -1230,7 +_,7 @@
             if (Platform.IsLinux)
             {
                 Process.Start(new ProcessStartInfo(folderPath)
-                { FileName = "open-folder", Arguments = folderPath, UseShellExecute = true, CreateNoWindow = true });
+                { FileName = "xdg-open", Arguments = folderPath, UseShellExecute = true, CreateNoWindow = true });
             }
             else
             {
@@ -2726,7 +_,7 @@
 
     public static string ReadEmbeddedResource(string path)
     {
-        using Stream stream = Assembly.GetExecutingAssembly().GetManifestResourceStream(path);
+        using Stream stream = Assembly.GetAssembly(typeof(Main)).GetManifestResourceStream(path);
         using StreamReader streamReader = new StreamReader(stream);
         return streamReader.ReadToEnd();
     }
