--- src/Terraria/Terraria/Utils.cs
+++ src/TerraAngel/Terraria/Utils.cs
@@ -1131,7 +_,8 @@
 
     public static Vector2 ScreenToWorldPosition(this Vector2 screenPosition)
     {
-        return Vector2.Transform(screenPosition * Main.UIScale, Matrix.Invert(Main.GameViewMatrix.TransformationMatrix)) + Main.screenPosition;
+        Matrix.Invert(Main.GameViewMatrix.TransformationMatrix, out var invertedTransformMatrix);
+        return Vector2.Transform(screenPosition * Main.UIScale, invertedTransformMatrix) + Main.screenPosition;
     }
 
     public static string PrettifyPercentDisplay(float percent, string originalFormat)
@@ -2681,7 +_,7 @@
 
     public static string ReadEmbeddedResource(string path)
     {
-        using Stream stream = Assembly.GetExecutingAssembly().GetManifestResourceStream(path);
+        using Stream stream = Assembly.GetAssembly(typeof(Main)).GetManifestResourceStream(path);
         using StreamReader streamReader = new StreamReader(stream);
         return streamReader.ReadToEnd();
     }
