--- src/Terraria/Terraria/WorldGen.cs
+++ src/TerraAngel/Terraria/WorldGen.cs
@@ -1062,18 +_,20 @@
                     continue;
                 }
 
-                Tile tile = new Tile(Main.tile[num6, num7]);
-                Tile tile2 = new Tile(Main.tile[num8, num9]);
-                Main.tile[num6, num7].active(tile2.active());
-                Main.tile[num6, num7].type = tile2.type;
-                Main.tile[num6, num7].color(tile2.color());
-                Main.tile[num6, num7].fullbrightBlock(tile2.fullbrightBlock());
-                Main.tile[num6, num7].invisibleBlock(tile2.invisibleBlock());
-                Main.tile[num8, num9].active(tile.active());
-                Main.tile[num8, num9].type = tile.type;
-                Main.tile[num8, num9].color(tile.color());
-                Main.tile[num8, num9].fullbrightBlock(tile.fullbrightBlock());
-                Main.tile[num8, num9].invisibleBlock(tile.invisibleBlock());
+                ref TileData tile = ref Main.tile.GetTileRef(num6, num7);
+                ref TileData tile2 = ref Main.tile.GetTileRef(num8, num9);
+                TileData tileCopy = tile;
+                TileData tile2Copy = tile2;
+                tile.active(tile2Copy.active());
+                tile.type = tile2Copy.type;
+                tile.color(tile2Copy.color());
+                tile.fullbrightBlock(tile2Copy.fullbrightBlock());
+                tile.invisibleBlock(tile2Copy.invisibleBlock());
+                tile2.active(tileCopy.active());
+                tile2.type = tileCopy.type;
+                tile2.color(tileCopy.color());
+                tile2.fullbrightBlock(tileCopy.fullbrightBlock());
+                tile2.invisibleBlock(tileCopy.invisibleBlock());
             }
 
             num3 = Main.maxTilesX / 2 / Variations.errorWorldAdjustment(1.0);
@@ -1158,18 +_,20 @@
                 {
                     for (int num18 = 0; num18 < num11; num18++)
                     {
-                        Tile tile3 = new Tile(Main.tile[num12 + num17, num13 + num18]);
-                        Tile tile4 = new Tile(Main.tile[num14 + num17, num15 + num18]);
-                        Main.tile[num12 + num17, num13 + num18].active(tile4.active());
-                        Main.tile[num12 + num17, num13 + num18].type = tile4.type;
-                        Main.tile[num12 + num17, num13 + num18].color(tile4.color());
-                        Main.tile[num12 + num17, num13 + num18].fullbrightBlock(tile4.fullbrightBlock());
-                        Main.tile[num12 + num17, num13 + num18].invisibleBlock(tile4.invisibleBlock());
-                        Main.tile[num14 + num17, num15 + num18].active(tile3.active());
-                        Main.tile[num14 + num17, num15 + num18].type = tile3.type;
-                        Main.tile[num14 + num17, num15 + num18].color(tile3.color());
-                        Main.tile[num14 + num17, num15 + num18].fullbrightBlock(tile3.fullbrightBlock());
-                        Main.tile[num14 + num17, num15 + num18].invisibleBlock(tile3.invisibleBlock());
+                        ref TileData tile3 = ref Main.tile.GetTileRef(num12 + num17, num13 + num18);
+                        ref TileData tile4 = ref Main.tile.GetTileRef(num14 + num17, num15 + num18);
+                        TileData tile3Copy = tile3;
+                        TileData tile4Copy = tile4;
+                        tile3.active(tile4Copy.active());
+                        tile3.type = tile4Copy.type;
+                        tile3.color(tile4Copy.color());
+                        tile3.fullbrightBlock(tile4Copy.fullbrightBlock());
+                        tile3.invisibleBlock(tile4Copy.invisibleBlock());
+                        tile4.active(tile3Copy.active());
+                        tile4.type = tile3Copy.type;
+                        tile4.color(tile3Copy.color());
+                        tile4.fullbrightBlock(tile3Copy.fullbrightBlock());
+                        tile4.invisibleBlock(tile3Copy.invisibleBlock());
                     }
                 }
             }
@@ -1209,7 +_,7 @@
                 int num22 = genRand.Next(5, 21);
                 int num23 = num20;
                 int num24 = num21;
-                Tile tile5 = new Tile(Main.tile[num20, num21]);
+                Tile tile5 = Main.tile[num20, num21];
                 for (int num25 = 0; num25 < num22; num25++)
                 {
                     num23 += (int)vector2D.X;
@@ -89341,8 +_,8 @@
                                     break;
                             }
 
-                            int num2 = tile2?.frameY ?? (-1);
-                            int num3 = tile?.frameY ?? (-1);
+                            int num2 = tile2.frameY;
+                            int num3 = tile.frameY;
                             bool flag = false;
                             switch (num)
                             {
@@ -89380,8 +_,8 @@
                                 up = num;
                             }
 
-                            int num4 = tile2?.frameY ?? (-1);
-                            int num5 = tile?.frameY ?? (-1);
+                            int num4 = tile2.frameY;
+                            int num5 = tile.frameY;
                             bool flag4 = false;
                             switch (num)
                             {
