--- src/Terraria/Terraria/GameContent/Drawing/TileDrawing.cs
+++ src/TerraAngel/Terraria/GameContent/Drawing/TileDrawing.cs
@@ -2,6 +_,7 @@
 using System.Collections.Generic;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
+using TerraAngel.Tools.Visuals;
 using Terraria.DataStructures;
 using Terraria.GameContent.Events;
 using Terraria.GameContent.Liquid;
@@ -136,11 +_,6 @@
             for (int j = firstTileX - 2; j < lastTileX + 2; j++)
             {
                 Tile tile = Main.tile[j, i];
-                if (tile == null)
-                {
-                    continue;
-                }
-
                 if (tile.active())
                 {
                     Main.instance.LoadTiles(tile.type);
@@ -401,11 +_,8 @@
             for (int j = firstTileX - 2; j < lastTileX + 2; j++)
             {
                 Tile tile = Main.tile[j, i];
-                if (tile != null)
-                {
                     Main.tileBatch.SetLayer(0u, 0);
                     DrawTile_LiquidBehindTile(solidLayer: false, waterStyleOverride, unscaledPosition, drawOffSet, j, i, tile);
-                }
             }
         }
 
@@ -454,16 +_,8 @@
         {
             for (int j = firstTileX - 2; j < lastTileX + 2; j++)
             {
-                Tile tile = Main.tile[j, i];
-                if (tile == null)
-                {
-                    tile = new Tile();
-                    Main.tile[j, i] = tile;
-                    Main.mapTime += 60;
-                }
-                else
-                {
-                    if (!tile.active() || IsTileDrawLayerSolid(tile.type) != solidLayer || (DebugOptions.ShowUnbreakableWall && tile.wall == 350))
+                ref TileData tile = ref Main.tile.GetTileRef(j, i);
+                if (!tile.active() || IsTileDrawLayerSolid(tile.type) != solidLayer || (DebugOptions.ShowUnbreakableWall && tile.wall == 350))
                     {
                         continue;
                     }
@@ -471,7 +_,7 @@
                     if (solidLayer)
                     {
                         Main.tileBatch.SetLayer(Layer_LiquidBehindTiles, 0);
-                        DrawTile_LiquidBehindTile(solidLayer, waterStyleOverride, unscaledPosition, drawOffSet, j, i, tile);
+                        DrawTile_LiquidBehindTile(solidLayer, waterStyleOverride, unscaledPosition, drawOffSet, j, i, ref tile);
                     }
 
                     Main.tileBatch.SetLayer(Layer_Tiles, 0);
@@ -663,7 +_,7 @@
                         case 375:
                         case 461:
                         case 709:
-                            EmitLiquidDrops(i, j, tile, type);
+                            EmitLiquidDrops(i, j, ref tile, type);
                             continue;
                         case 491:
                             if (frameX == 18 && frameY == 18)
@@ -690,7 +_,7 @@
                             AddSpecialPoint(j, i, TileCounterType.AnyDirectionalGrass);
                             continue;
                         default:
-                            if (ShouldSwayInWind(j, i, tile))
+                            if (ShouldSwayInWind(j, i, ref tile))
                             {
                                 AddSpecialPoint(j, i, TileCounterType.WindyGrass);
                                 continue;
@@ -699,8 +_,103 @@
                             break;
                     }
 
+                    if (LightingModifierTool.LightingModifierToolCache?.FullBright ?? false)
+                    {
+                        if (tile.type == TileID.MinecartTrack)
+                        {
+                            DrawSingleTile(unscaledPosition, drawOffSet, j, i);
+                            continue;
+                        }
+                        Color tileColor = new Color(new Vector3(LightingModifierTool.LightingModifierToolCache.Brightness));
+                        ushort typeCache = tile.type;
+                        short tileFrameX = tile.frameX;
+                        short tileFrameY = tile.frameY;
+                        GetTileDrawData(j, i, ref tile, typeCache, ref tileFrameX, ref tileFrameY, out var tileWidth, out var tileHeight, out var tileTop, out var halfBrickHeight, out var addFrX, out var addFrY, out var tileSpriteEffect, out var glowTexture, out var glowSourceRect, out var glowColor);
+                        Texture2D texture = GetTileDrawTexture(ref tile, j, i);
+                        if (texture == null)
+                        {
+                            continue;
+                        }
+                        Rectangle normalTileRect = new Rectangle(tileFrameX + addFrX, tileFrameY + addFrY, tileWidth, tileHeight - halfBrickHeight);
+                        Vector2 normalTilePosition = new Vector2((float)(j * 16 - (int)unscaledPosition.X) - ((float)tileWidth - 16f) / 2f, i * 16 - (int)unscaledPosition.Y + tileTop + halfBrickHeight) + drawOffSet;
+
+                        int slope = tile.slope();
+                        if (slope > 0)
+                        {
+                            if (TileID.Sets.Platforms[typeCache])
+                            {
+                                Main.spriteBatch.Draw(texture, normalTilePosition, normalTileRect, tileColor, 0f, _zero, 1f, tileSpriteEffect, 0f);
+                                if (slope == 1 && Main.tile[j + 1, i + 1].active() && Main.tileSolid[Main.tile[j + 1, i + 1].type] && Main.tile[j + 1, i + 1].slope() != 2 && !Main.tile[j + 1, i + 1].halfBrick() && (!Main.tile[j, i + 1].active() || (Main.tile[j, i + 1].blockType() != 0 && Main.tile[j, i + 1].blockType() != 5) || !TileID.Sets.BlocksStairs[Main.tile[j, i + 1].type]))
+                                {
+                                    Rectangle rectValue = new Rectangle(198, tileFrameY, 16, 16);
+                                    if (TileID.Sets.Platforms[Main.tile[j + 1, i + 1].type] && Main.tile[j + 1, i + 1].slope() == 0)
+                                    {
+                                        rectValue.X = 324;
+                                    }
+
+                                    Main.spriteBatch.Draw(texture, normalTilePosition + new Vector2(0f, 16f), rectValue, tileColor, 0f, _zero, 1f, tileSpriteEffect, 0f);
+                                }
+                                else if (slope == 2 && Main.tile[j - 1, i + 1].active() && Main.tileSolid[Main.tile[j - 1, i + 1].type] && Main.tile[j - 1, i + 1].slope() != 1 && !Main.tile[j - 1, i + 1].halfBrick() && (!Main.tile[j, i + 1].active() || (Main.tile[j, i + 1].blockType() != 0 && Main.tile[j, i + 1].blockType() != 4) || !TileID.Sets.BlocksStairs[Main.tile[j, i + 1].type]))
+                                {
+                                    Rectangle value2 = new Rectangle(162, tileFrameY, 16, 16);
+                                    if (TileID.Sets.Platforms[Main.tile[j - 1, i + 1].type] && Main.tile[j - 1, i + 1].slope() == 0)
+                                    {
+                                        value2.X = 306;
+                                    }
+
+                                    Main.spriteBatch.Draw(texture, normalTilePosition + new Vector2(0f, 16f), value2, tileColor, 0f, _zero, 1f, tileSpriteEffect, 0f);
+                                }
+
+                                continue;
+                            }
+                            if (TileID.Sets.HasSlopeFrames[typeCache])
+                            {
+                                Main.spriteBatch.Draw(texture, normalTilePosition, new Rectangle(tileFrameX + addFrX, tileFrameY + addFrY, 16, 16), tileColor, 0f, _zero, 1f, tileSpriteEffect, 0f);
+                                continue;
+                            }
+
+                            int l_num = slope;
+                            int l_num2 = 2;
+                            for (int l_i = 0; l_i < 8; l_i++)
+                            {
+                                int l_num3 = l_i * -2;
+                                int l_num4 = 16 - l_i * 2;
+                                int l_num5 = 16 - l_num4;
+                                int l_num6;
+                                switch (l_num)
+                                {
+                                    case 1:
+                                        l_num3 = 0;
+                                        l_num6 = l_i * 2;
+                                        l_num4 = 14 - l_i * 2;
+                                        l_num5 = 0;
+                                        break;
+                                    case 2:
+                                        l_num3 = 0;
+                                        l_num6 = 16 - l_i * 2 - 2;
+                                        l_num4 = 14 - l_i * 2;
+                                        l_num5 = 0;
+                                        break;
+                                    case 3:
+                                        l_num6 = l_i * 2;
+                                        break;
+                                    default:
+                                        l_num6 = 16 - l_i * 2 - 2;
+                                        break;
+                                }
+
+                                Main.spriteBatch.Draw(texture, normalTilePosition + new Vector2(l_num6, l_i * l_num2 + l_num3), new Rectangle(tileFrameX + addFrX + l_num6, tileFrameY + addFrY + l_num5, l_num2, l_num4), tileColor, 0f, _zero, 1f, tileSpriteEffect, 0f);
+                            }
+
+                            int l_num7 = ((l_num <= 2) ? 14 : 0);
+                            Main.spriteBatch.Draw(texture, normalTilePosition + new Vector2(0f, l_num7), new Rectangle(tileFrameX + addFrX, tileFrameY + addFrY + l_num7, 16, 2), tileColor, 0f, _zero, 1f, tileSpriteEffect, 0f);
+                            continue;
+                        }
+                        Main.spriteBatch.Draw(texture, normalTilePosition, normalTileRect, tileColor, 0f, _zero, 1f, tileSpriteEffect, 0f);
+                        continue;
+                    }
+
                     DrawSingleTile(unscaledPosition, drawOffSet, j, i);
-                }
             }
         }
 
@@ -1207,7 +_,7 @@
         }
         else
         {
-            DrawBasicTile(screenPosition, screenOffset, tileX, tileY, tileDrawInfo, rectangle, vector);
+            DrawBasicTile(screenPosition, screenOffset, tileX, tileY, ref tileDrawInfo, rectangle, vector);
         }
 
         if (Main.tileGlowMask[tileDrawInfo.tileCache.type] != -1)
@@ -1485,7 +_,9 @@
         return true;
     }
 
+    public Texture2D GetTileDrawTexture(Tile tile, int tileX, int tileY) => GetTileDrawTexture(ref tile.RefData, tileX, tileY);
+
-    public Texture2D GetTileDrawTexture(Tile tile, int tileX, int tileY)
+    public Texture2D GetTileDrawTexture(ref TileData tile, int tileX, int tileY)
     {
         TilePaintSystemV2.TileVariationkey tileVariationkey = default(TilePaintSystemV2.TileVariationkey);
         tileVariationkey.TileType = tile.type;
@@ -1558,7 +_,7 @@
         return TextureAssets.CageTop[key.CageStyle].Value;
     }
 
-    private void DrawBasicTile(Vector2 screenPosition, Vector2 screenOffset, int tileX, int tileY, TileDrawInfo drawData, Rectangle normalTileRect, Vector2 normalTilePosition)
+    private void DrawBasicTile(Vector2 screenPosition, Vector2 screenOffset, int tileX, int tileY, ref TileDrawInfo drawData, Rectangle normalTileRect, Vector2 normalTilePosition)
     {
         bool flag = !TileID.Sets.DontDrawTileSliced[drawData.tileCache.type];
         if (drawData.typeCache == 380 || TileID.Sets.Platforms[drawData.typeCache])
@@ -1851,7 +_,7 @@
 
     public static int GetTreeVariant(int x, int y)
     {
-        if (Main.tile[x, y] == null || !Main.tile[x, y].active())
+        if (!Main.tile[x, y].active())
         {
             return -1;
         }
@@ -3906,7 +_,7 @@
     private void DrawXmasTree(Vector2 screenPosition, Vector2 screenOffset, int tileX, int tileY, TileDrawInfo drawData)
     {
         int num = 2;
-        if (tileY - drawData.tileFrameY > 0 && drawData.tileFrameY == 7 && Main.tile[tileX, tileY - drawData.tileFrameY] != null)
+        if (tileY - drawData.tileFrameY > 0 && drawData.tileFrameY == 7)
         {
             drawData.tileTop -= 16 * drawData.tileFrameY;
             drawData.tileFrameX = Main.tile[tileX, tileY - drawData.tileFrameY].frameX;
@@ -4037,12 +_,9 @@
         if (WorldGen.IsRope(tileX, tileY, out topRopeY, out bottomRopeY))
         {
             Tile tile = Main.tile[tileX, topRopeY];
-            if (tile != null)
-            {
                 int y = (tileY + tileX) % 3 * 18;
                 Texture2D tileDrawTexture = GetTileDrawTexture(tile, tileX, tileY);
                 Main.tileBatch.Draw(tileDrawTexture, new Vector2(tileX * 16 - (int)screenPosition.X, tileY * 16 - (int)screenPosition.Y) + screenOffset, new Rectangle(90, y, 16, 16), drawData.tileLight, default(Vector2), 1f, drawData.tileSpriteEffect);
-            }
         }
     }
 
@@ -4089,35 +_,14 @@
         }
     }
 
+    private void DrawTile_LiquidBehindTile(bool solidLayer, int waterStyleOverride, Vector2 screenPosition, Vector2 screenOffset, int tileX, int tileY, ref TileData tileCache) => DrawTile_LiquidBehindTile(solidLayer, waterStyleOverride, screenPosition, screenOffset, tileX, tileY, new Tile(ref tileCache));
+
     private void DrawTile_LiquidBehindTile(bool solidLayer, int waterStyleOverride, Vector2 screenPosition, Vector2 screenOffset, int tileX, int tileY, Tile tileCache)
     {
         Tile tile = Main.tile[tileX + 1, tileY];
         Tile tile2 = Main.tile[tileX - 1, tileY];
         Tile tile3 = Main.tile[tileX, tileY - 1];
         Tile tile4 = Main.tile[tileX, tileY + 1];
-        if (tile == null)
-        {
-            tile = new Tile();
-            Main.tile[tileX + 1, tileY] = tile;
-        }
-
-        if (tile2 == null)
-        {
-            tile2 = new Tile();
-            Main.tile[tileX - 1, tileY] = tile2;
-        }
-
-        if (tile3 == null)
-        {
-            tile3 = new Tile();
-            Main.tile[tileX, tileY - 1] = tile3;
-        }
-
-        if (tile4 == null)
-        {
-            tile4 = new Tile();
-            Main.tile[tileX, tileY + 1] = tile4;
-        }
 
         if (tile.type == 379)
         {
@@ -4867,6 +_,8 @@
         return WorldGen.InAPlaceWithWind(x, y, width, height);
     }
 
+    private void GetTileDrawData(int x, int y, ref TileData tileCache, ushort typeCache, ref short tileFrameX, ref short tileFrameY, out int tileWidth, out int tileHeight, out int tileTop, out int halfBrickHeight, out int addFrX, out int addFrY, out SpriteEffects tileSpriteEffect, out Texture2D glowTexture, out Rectangle glowSourceRect, out Color glowColor) => GetTileDrawData(x, y, new Tile(ref tileCache), typeCache, ref tileFrameX, ref tileFrameY, out tileWidth, out tileHeight, out tileTop, out halfBrickHeight, out addFrX, out addFrY, out tileSpriteEffect, out glowTexture, out glowSourceRect, out glowColor);
+
     private void GetTileDrawData(int x, int y, Tile tileCache, ushort typeCache, ref short tileFrameX, ref short tileFrameY, out int tileWidth, out int tileHeight, out int tileTop, out int halfBrickHeight, out int addFrX, out int addFrY, out SpriteEffects tileSpriteEffect, out Texture2D glowTexture, out Rectangle glowSourceRect, out Color glowColor)
     {
         tileTop = 0;
@@ -7119,11 +_,6 @@
     private bool IsWindBlocked(int x, int y)
     {
         Tile tile = Main.tile[x, y];
-        if (tile == null)
-        {
-            return true;
-        }
-
         if (tile.wall > 0 && !WallID.Sets.AllowsWind[tile.wall])
         {
             return true;
@@ -7534,7 +_,7 @@
                 bool flag2 = _rand.Next(600) == 0;
                 if (!flag2)
                 {
-                    _windGrid.GetWindTime(i, j, 8, out var windTimeLeft, out var directionX, out directionX);
+                    _windGrid.GetWindTime(i, j, 8, out var windTimeLeft, out var directionX, out _);
                     flag2 = windTimeLeft > 0 && _rand.Next(48) == 0;
                 }
 
@@ -8307,7 +_,7 @@
         }
     }
 
-    private void EmitLiquidDrops(int j, int i, Tile tileCache, ushort typeCache)
+    private void EmitLiquidDrops(int j, int i, ref TileData tileCache, ushort typeCache)
     {
         int num = 60;
         switch (typeCache)
@@ -8435,7 +_,7 @@
         return num2 * lerpValue;
     }
 
-    private bool ShouldSwayInWind(int x, int y, Tile tileCache)
+    private bool ShouldSwayInWind(int x, int y, ref TileData tileCache)
     {
         if (!Main.SettingsEnabled_TilesSwayInWind)
         {
@@ -9926,11 +_,6 @@
         for (int i = startY; i < Main.maxTilesY - 10; i++)
         {
             Tile tile = Main.tile[x, i];
-            if (tile == null)
-            {
-                break;
-            }
-
             ushort type = tile.type;
             if (!tile.active() || !TileID.Sets.VineThreads[type])
             {
@@ -10041,8 +_,6 @@
         for (int num6 = startY; num6 > 10; num6--)
         {
             Tile tile = Main.tile[x, num6];
-            if (tile != null)
-            {
                 ushort type = tile.type;
                 if (!tile.active() || !TileID.Sets.ReverseVineThreads[type])
                 {
@@ -10093,7 +_,6 @@
                 }
 
                 vector += (num7 - (float)Math.PI / 2f).ToRotationVector2() * 16f;
-            }
         }
     }
 
