--- src/Terraria/Terraria/GameContent/Drawing/WallDrawing.cs
+++ src/TerraAngel/Terraria/GameContent/Drawing/WallDrawing.cs
@@ -1,5 +_,6 @@
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
+using TerraAngel.Vectors;
 using Terraria.GameContent.Liquid;
 using Terraria.Graphics;
 using Terraria.ID;
@@ -10,7 +_,7 @@
 {
     public static bool QuickPaintLookup = true;
     private static VertexColors _glowPaintColors = new VertexColors(Color.White);
-    private Tile[,] _tileArray;
+    private NativeTileMap _tileArray;
     private TilePaintSystemV2 _paintSystem;
     private bool _shouldShowInvisibleWalls;
     private TilePaintSystemV2.WallVariationKey _lastPaintLookupKey;
@@ -52,6 +_,7 @@
 
         float gfxQuality = Main.gfxQuality;
         Vector2 screenPosition = Main.screenPosition;
+        Vector2i intScreenPosition = screenPosition;
         int[] wallBlend = Main.wallBlend;
         _tileArray = Main.tile;
         int num = (int)(120f * (1f - gfxQuality) + 40f * gfxQuality);
@@ -139,12 +_,12 @@
                         }
                     }
 
-                    Main.tileBatch.Draw(wallDrawTexture, new Vector2(j * 16 - (int)screenPosition.X - 8, i * 16 - (int)screenPosition.Y - 8) + drawOffSet, value, vertices, Vector2.Zero, 1f, SpriteEffects.None);
+                    Main.tileBatch.Draw(wallDrawTexture, new Vector2(j * 16 - intScreenPosition.X - 8, i * 16 - intScreenPosition.Y - 8) + drawOffSet, value, vertices, Vector2.Zero, 1f, SpriteEffects.None);
                     if (tile.wall == 347)
                     {
                         Texture2D value2 = TextureAssets.GlowMask[361].Value;
                         LiquidRenderer.SetShimmerVertexColors_Sparkle(ref vertices, 0.7f, j, i, top: true);
-                        Main.tileBatch.Draw(value2, new Vector2(j * 16 - (int)screenPosition.X - 8, i * 16 - (int)screenPosition.Y - 8) + drawOffSet, value, vertices, Vector2.Zero, 1f, SpriteEffects.None);
+                        Main.tileBatch.Draw(value2, new Vector2(j * 16 - intScreenPosition.X - 8, i * 16 - intScreenPosition.Y - 8) + drawOffSet, value, vertices, Vector2.Zero, 1f, SpriteEffects.None);
                     }
                 }
                 else
@@ -160,12 +_,12 @@
                         color2 = Color.Lerp(color2, Color.White, 0.5f);
                     }
 
-                    Main.tileBatch.Draw(wallDrawTexture, new Vector2(j * 16 - (int)screenPosition.X - 8, i * 16 - (int)screenPosition.Y - 8) + drawOffSet, value, color2, Vector2.Zero, 1f, SpriteEffects.None);
+                    Main.tileBatch.Draw(wallDrawTexture, new Vector2(j * 16 - intScreenPosition.X - 8, i * 16 - intScreenPosition.Y - 8) + drawOffSet, value, color2, Vector2.Zero, 1f, SpriteEffects.None);
                     if (tile.wall == 347)
                     {
                         Texture2D value3 = TextureAssets.GlowMask[361].Value;
                         Color color3 = LiquidRenderer.GetShimmerGlitterColor(top: true, j, i) * 0.7f;
-                        Main.tileBatch.Draw(value3, new Vector2(j * 16 - (int)screenPosition.X - 8, i * 16 - (int)screenPosition.Y - 8) + drawOffSet, value, color3, Vector2.Zero, 1f, SpriteEffects.None);
+                        Main.tileBatch.Draw(value3, new Vector2(j * 16 - intScreenPosition.X - 8, i * 16 - intScreenPosition.Y - 8) + drawOffSet, value, color3, Vector2.Zero, 1f, SpriteEffects.None);
                     }
                 }
 
@@ -177,22 +_,22 @@
                     bool flag3 = _tileArray[j, i + 1].wall > 0 && wallBlend[_tileArray[j, i + 1].wall] != wallBlend[tile.wall];
                     if (num7)
                     {
-                        DrawOutline(TextureAssets.WallOutline.Value, new Vector2(j * 16 - (int)screenPosition.X, i * 16 - (int)screenPosition.Y) + drawOffSet, new Rectangle(0, 0, 2, 16), color);
+                        DrawOutline(TextureAssets.WallOutline.Value, new Vector2(j * 16 - intScreenPosition.X, i * 16 - intScreenPosition.Y) + drawOffSet, new Rectangle(0, 0, 2, 16), color);
                     }
 
                     if (flag)
                     {
-                        DrawOutline(TextureAssets.WallOutline.Value, new Vector2(j * 16 - (int)screenPosition.X + 14, i * 16 - (int)screenPosition.Y) + drawOffSet, new Rectangle(14, 0, 2, 16), color);
+                        DrawOutline(TextureAssets.WallOutline.Value, new Vector2(j * 16 - intScreenPosition.X + 14, i * 16 - intScreenPosition.Y) + drawOffSet, new Rectangle(14, 0, 2, 16), color);
                     }
 
                     if (flag2)
                     {
-                        DrawOutline(TextureAssets.WallOutline.Value, new Vector2(j * 16 - (int)screenPosition.X, i * 16 - (int)screenPosition.Y) + drawOffSet, new Rectangle(0, 0, 16, 2), color);
+                        DrawOutline(TextureAssets.WallOutline.Value, new Vector2(j * 16 - intScreenPosition.X, i * 16 - intScreenPosition.Y) + drawOffSet, new Rectangle(0, 0, 16, 2), color);
                     }
 
                     if (flag3)
                     {
-                        DrawOutline(TextureAssets.WallOutline.Value, new Vector2(j * 16 - (int)screenPosition.X, i * 16 - (int)screenPosition.Y + 14) + drawOffSet, new Rectangle(0, 14, 16, 2), color);
+                        DrawOutline(TextureAssets.WallOutline.Value, new Vector2(j * 16 - intScreenPosition.X, i * 16 - intScreenPosition.Y + 14) + drawOffSet, new Rectangle(0, 14, 16, 2), color);
                     }
                 }
             }
