--- src/Terraria/Terraria/Net/Sockets/TcpSocket.cs
+++ src/TerraAngel/Terraria/Net/Sockets/TcpSocket.cs
@@ -13,6 +_,7 @@
     private SocketConnectionAccepted _listenerCallback;
     private RemoteAddress _remoteAddress;
     private bool _isListening;
+    private NetworkStream networkStream;
     private DebugNetworkStream _debugStream;
     private DebugNetworkStream GetStream()
     {
@@ -58,6 +_,7 @@
         TcpAddress tcpAddress = (TcpAddress)address;
         _connection.Connect(tcpAddress.Address, tcpAddress.Port);
         _remoteAddress = address;
+        networkStream = _connection.GetStream();
     }
 
     private void ReadCallback(IAsyncResult result)
@@ -65,7 +_,8 @@
         try
         {
             Tuple<SocketReceiveCallback, object> tuple = (Tuple<SocketReceiveCallback, object>)result.AsyncState;
-            tuple.Item1(tuple.Item2, GetStream().EndRead(result));
+            int messageLength = networkStream.EndRead(result);
+            tuple.Item1(tuple.Item2, messageLength);
         }
         catch (ObjectDisposedException)
         {
@@ -76,9 +_,12 @@
     private void SendCallback(IAsyncResult result)
     {
         Tuple<SocketSendCallback, object> tuple;
+        int messageLength = 0;
         if (Platform.IsWindows)
         {
+            object[] data = (object[])result.AsyncState!;
-            tuple = (Tuple<SocketSendCallback, object>)result.AsyncState;
+            tuple = (Tuple<SocketSendCallback, object>)data[0];
+            messageLength = (int)data[1];
         }
         else
         {
@@ -89,7 +_,7 @@
 
         try
         {
-            GetStream().EndWrite(result);
+            networkStream.EndWrite(result);
             tuple.Item1(tuple.Item2);
         }
         catch (Exception)
@@ -103,17 +_,17 @@
         if (!Platform.IsWindows)
         {
             byte[] array = LegacyNetBufferPool.RequestBuffer(data, offset, size);
-            GetStream().BeginWrite(array, 0, size, SendCallback, new object[2] { new Tuple<SocketSendCallback, object>(callback, state), array });
+            networkStream.BeginWrite(array, 0, size, SendCallback, new object[] { new Tuple<SocketSendCallback, object>(callback, state), array });
         }
         else
         {
-            GetStream().BeginWrite(data, 0, size, SendCallback, new Tuple<SocketSendCallback, object>(callback, state));
+            networkStream.BeginWrite(data, 0, size, SendCallback, new object[] { new Tuple<SocketSendCallback, object>(callback, state), size });
         }
     }
 
     void ISocket.AsyncReceive(byte[] data, int offset, int size, SocketReceiveCallback callback, object state)
     {
-        GetStream().BeginRead(data, offset, size, ReadCallback, new Tuple<SocketReceiveCallback, object>(callback, state));
+        networkStream.BeginRead(data, offset, size, ReadCallback, new Tuple<SocketReceiveCallback, object>(callback, state));
     }
 
     bool ISocket.IsDataAvailable()
@@ -123,7 +_,7 @@
             return false;
         }
 
-        return GetStream().DataAvailable;
+        return networkStream.DataAvailable;
     }
 
     RemoteAddress ISocket.GetRemoteAddress()
