--- src/Terraria/Terraria/Testing/QuickLoad.cs
+++ src/TerraAngel/Terraria/Testing/QuickLoad.cs
@@ -57,7 +_,6 @@
                 Rectangle value = WindowedBounds.Value;
                 if (Platform.Get<IWindowService>().IsSizeable(Main.instance.Window))
                 {
-                    Main.SetResolution(value.Width, value.Height);
                     Platform.Get<IWindowService>().SetPosition(Main.instance.Window, value.X, value.Y);
                 }
             }
