--- src/Terraria/Terraria/Map/MapHelper.cs
+++ src/TerraAngel/Terraria/Map/MapHelper.cs
@@ -1860,22 +_,20 @@
         return mapType >= skyPosition;
     }
 
+    // patch context
     public static MapTile CreateMapTile(int i, int j, byte Light, int backgroundOverride = 0)
     {
-        Tile tile = Main.tile[i, j];
-        if (tile == null)
-        {
-            return default(MapTile);
-        }
+        // patch context
+        ref TileData tile = ref Main.tile.GetTileRef(i, j);
 
         int newColor = 0;
         int newLight = Light;
         int baseType = 0;
         int baseOption = 0;
-        GetTileType(i, j, tile, ref newColor, ref newLight, ref baseType, ref baseOption);
+        GetTileType(i, j, ref tile, ref newColor, ref newLight, ref baseType, ref baseOption);
         if (baseType == 0)
         {
-            GetWallType(i, j, tile, ref newColor, ref newLight, ref baseType, ref baseOption);
+            GetWallType(i, j, ref tile, ref newColor, ref newLight, ref baseType, ref baseOption);
         }
 
         if (baseType == 0)
@@ -1888,7 +_,7 @@
         return MapTile.Create((ushort)(baseType + baseOption), (byte)newLight, (byte)newColor);
     }
 
-    private static void GetTileType(int i, int j, Tile tileCache, ref int newColor, ref int newLight, ref int baseType, ref int baseOption)
+    private static void GetTileType(int i, int j, ref TileData tileCache, ref int newColor, ref int newLight, ref int baseType, ref int baseOption)
     {
         if ((DebugOptions.ShowUnbreakableWall && tileCache.wall == 350) || !tileCache.active())
         {
@@ -1946,7 +_,7 @@
 
         if (baseType != 0)
         {
-            GetTileBaseOption(i, j, num, tileCache, ref baseOption);
+            GetTileBaseOption(i, j, num, ref tileCache, ref baseOption);
             if (num == 160)
             {
                 newColor = 0;
@@ -1958,7 +_,7 @@
         }
     }
 
-    private static void GetWallType(int i, int j, Tile tileCache, ref int newColor, ref int newLight, ref int baseType, ref int baseOption)
+    private static void GetWallType(int i, int j, ref TileData tileCache, ref int newColor, ref int newLight, ref int baseType, ref int baseOption)
     {
         bool flag = tileCache.invisibleWall();
         if (tileCache.wall > 0 && tileCache.fullbrightWall() && !flag)
@@ -2061,7 +_,7 @@
         return skyPosition + num;
     }
 
-    public static void GetTileBaseOption(int x, int y, int tileType, Tile tileCache, ref int baseOption)
+    public static void GetTileBaseOption(int x, int y, int tileType, ref TileData tileCache, ref int baseOption)
     {
         switch (tileType)
         {
