--- src/Terraria/Terraria/Graphics/VertexStrip.cs
+++ src/TerraAngel/Terraria/Graphics/VertexStrip.cs
@@ -92,7 +_,7 @@
             Vector2 target = vector;
             for (float num7 = num6; num7 < 1f; num7 += num6)
             {
-                Vector2 vector3 = Vector2.CatmullRom(value, vector, vector2, value2, num7);
+                Vector2 vector3 = VectorExtensions.CatmullRom(value, vector, vector2, value2, num7);
                 float num8 = MathHelper.WrapAngle(vector3.DirectionTo(target).ToRotation());
                 if (float.IsNaN(num8))
                 {
