--- src/Terraria/Terraria/GameContent/Creative/ItemsSacrificedUnlocksTracker.cs
+++ src/TerraAngel/Terraria/GameContent/Creative/ItemsSacrificedUnlocksTracker.cs
@@ -86,6 +_,9 @@
 
     public bool IsFullyResearched(int itemId)
     {
+        if (ClientConfig.Settings.ForceAllCreativeUnlocks && !ItemID.Sets.Deprecated[itemId])
+            return true;
+
         if (TryGetSacrificeNumbers(itemId, out var amountWeHave, out var amountNeededTotal))
         {
             return amountWeHave >= amountNeededTotal;
