--- src/Terraria/Terraria/Social/Steam/NetClientSocialModule.cs
+++ src/TerraAngel/Terraria/Social/Steam/NetClientSocialModule.cs
@@ -3,7 +_,6 @@
 using Terraria.Localization;
 using Terraria.Net;
 using Terraria.Net.Sockets;
-using Terraria.Social.WeGame;
 
 namespace Terraria.Social.Steam;
 public class NetClientSocialModule : NetSocialModule
@@ -52,7 +_,6 @@
 
     public override void LaunchLocalServer(Process process, ServerMode mode)
     {
-        WeGameHelper.WriteDebugString("LaunchLocalServer");
         if (_lobby.State != 0)
         {
             _lobby.Leave();
@@ -146,7 +_,6 @@
 
     private void OnLobbyJoinRequest(GameLobbyJoinRequested_t result)
     {
-        WeGameHelper.WriteDebugString(" OnLobbyJoinRequest");
         if (_lobby.State != 0)
         {
             _lobby.Leave();
@@ -163,7 +_,6 @@
 
     private void OnLobbyEntered(LobbyEnter_t result, bool failure)
     {
-        WeGameHelper.WriteDebugString(" OnLobbyEntered");
         SteamNetworking.AllowP2PPacketRelay(bAllow: true);
         SendAuthTicket(_lobby.Owner);
         int num = 0;
@@ -210,12 +_,11 @@
 
     private void SendAuthTicket(CSteamID address)
     {
-        WeGameHelper.WriteDebugString(" SendAuthTicket");
         if (_authTicket == HAuthTicket.Invalid)
         {
             SteamNetworkingIdentity pSteamNetworkingIdentity = default(SteamNetworkingIdentity);
             pSteamNetworkingIdentity.SetSteamID(address);
-            _authTicket = SteamUser.GetAuthSessionTicket(_authData, _authData.Length, out _authDataLength, ref pSteamNetworkingIdentity);
+            _authTicket = SteamUser.GetAuthSessionTicket(_authData, _authData.Length, out _authDataLength);
         }
 
         int num = (int)(_authDataLength + 3);
@@ -256,13 +_,11 @@
 
     private void OnSessionConnectFail(P2PSessionConnectFail_t result)
     {
-        WeGameHelper.WriteDebugString(" OnSessionConnectFail");
         Close(result.m_steamIDRemote);
     }
 
     private void OnP2PSessionRequest(P2PSessionRequest_t result)
     {
-        WeGameHelper.WriteDebugString(" OnP2PSessionRequest");
         CSteamID steamIDRemote = result.m_steamIDRemote;
         if (_connectionStateMap.ContainsKey(steamIDRemote) && _connectionStateMap[steamIDRemote] != 0)
         {
