--- src/Terraria/Terraria/Utilities/PlatformUtilities.cs
+++ src/TerraAngel/Terraria/Utilities/PlatformUtilities.cs
@@ -27,18 +_,18 @@
 
     public static void SavePng(Stream stream, int width, int height, int imgWidth, int imgHeight, byte[] data)
     {
-        IntPtr surface = INTERNAL_getScaledSurface(data, width, height, imgWidth, imgHeight);
+        // IntPtr surface = INTERNAL_getScaledSurface(data, width, height, imgWidth, imgHeight);
-        byte[] array = new byte[width * height * 4 + 41 + 57 + 256];
+        // byte[] array = new byte[width * height * 4 + 41 + 57 + 256];
-        IntPtr intPtr = Marshal.AllocHGlobal(array.Length);
+        // IntPtr intPtr = Marshal.AllocHGlobal(array.Length);
-        IntPtr dst = SDL.SDL_RWFromMem(intPtr, array.Length);
+        // IntPtr dst = SDL.SDL_RWFromMem(intPtr, array.Length);
-        SDL_image.IMG_SavePNG_RW(surface, dst, 1);
+        // SDL_image.IMG_SavePNG_RW(surface, dst, 1);
-        SDL.SDL_FreeSurface(surface);
+        // SDL.SDL_FreeSurface(surface);
-        Marshal.Copy(intPtr, array, 0, array.Length);
+        // Marshal.Copy(intPtr, array, 0, array.Length);
-        Marshal.FreeHGlobal(intPtr);
+        // Marshal.FreeHGlobal(intPtr);
-        _ = array[33];
+        // _ = array[33];
-        _ = array[34];
+        // _ = array[34];
-        _ = array[35];
+        // _ = array[35];
-        _ = array[36];
+        // _ = array[36];
-        stream.Write(array, 0, array.Length);
+        // stream.Write(array, 0, array.Length);
     }
 }
