--- src/Terraria/Terraria/Graphics/Light/LightingEngine.cs
+++ src/TerraAngel/Terraria/Graphics/Light/LightingEngine.cs
@@ -37,6 +_,11 @@
     private LightMap _workingLightMap = new LightMap();
     private Rectangle _workingProcessedArea;
     private EngineState _state;
+    public int BlurPassCount
+    {
+        get => _activeLightMap.BlurPassCount;
+        set => _activeLightMap.BlurPassCount = (_workingLightMap.BlurPassCount = value);
+    }
     public void AddLight(int x, int y, Vector3 color)
     {
         _perFrameLights.Add(new PerFrameLight(new Point(x, y), color));
@@ -196,7 +_,7 @@
         workingLightMap.LightDecayThroughAir *= Main.SceneState.airLightDecay;
         workingLightMap.LightDecayThroughSolid *= Main.SceneState.solidLightDecay;
     }
-
+    
     private void ApplyPerFrameLights()
     {
         for (int i = 0; i < _perFrameLights.Count; i++)
@@ -206,13 +_,17 @@
             {
                 Vector3 value = _perFrameLights[i].Color;
                 Vector3 value2 = _workingLightMap[position.X - _workingProcessedArea.X, position.Y - _workingProcessedArea.Y];
-                Vector3.Max(ref value2, ref value, out value);
+                value = Vector3.Max(value2, value);
                 _workingLightMap[position.X - _workingProcessedArea.X, position.Y - _workingProcessedArea.Y] = value;
             }
         }
 
         if (!CaptureManager.Instance.IsCapturing)
         {
+        if (Lighting.AbleToProcessPerFrameLights)
+        {
+            Lighting.AbleToProcessPerFrameLights = false;
+        }
             _perFrameLights.Clear();
         }
     }
