--- src/Terraria/Terraria/Audio/LegacyAudioSystem.cs
+++ src/TerraAngel/Terraria/Audio/LegacyAudioSystem.cs
@@ -119,7 +_,7 @@
 
     public ReadOnlyCollection<RendererDetail> VeryExpensiveWayToGetAllAudioRenderers()
     {
-        AudioEngine audioEngine = new AudioEngine("Content\\TerrariaMusic.xgs");
+        AudioEngine audioEngine = new AudioEngine($"{Main.instance.Content.RootDirectory}\\TerrariaMusic.xgs");
         ReadOnlyCollection<RendererDetail> rendererDetails = audioEngine.RendererDetails;
         audioEngine.Dispose();
         return rendererDetails;
@@ -132,10 +_,11 @@
 
     public void Init(string renderer = "")
     {
-        Engine = new AudioEngine("Content\\TerrariaMusic.xgs", TimeSpan.FromMilliseconds(250.0), renderer);
+        Engine = new AudioEngine($"{Main.instance.Content.RootDirectory}\\TerrariaMusic.xgs", TimeSpan.FromMilliseconds(250.0), renderer);
-        SoundBank = new SoundBank(Engine, "Content\\Sound Bank.xsb");
+        SoundBank = new SoundBank(Engine, $"{Main.instance.Content.RootDirectory}\\Sound Bank.xsb");
         Engine.Update();
-        WaveBank = new WaveBank(Engine, "Content\\Wave Bank.xwb", 0, 512);
+        if (File.Exists("LContent\\Wave Bank.xwb")) WaveBank = new WaveBank(Engine, $"LContent\\Wave Bank.xwb", 0, 512);
+        else WaveBank = new WaveBank(Engine, $"{Main.instance.Content.RootDirectory}\\Wave Bank.xwb", 0, 512);
         Engine.Update();
         AudioTracks = new IAudioTrack[Main.maxMusic];
         TrackNamesByIndex = new Dictionary<int, string>();
