--- src/Terraria/Terraria/UI/FancyErrorPrinter.cs
+++ src/TerraAngel/Terraria/UI/FancyErrorPrinter.cs
@@ -141,7 +_,7 @@
     {
         if (!Main.dedServ)
         {
-            MessageBox.Show(preparedMessage, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
+            SDL3.SDL.SDL_ShowSimpleMessageBox(SDL3.SDL.SDL_MessageBoxFlags.SDL_MESSAGEBOX_ERROR, "Terraria: Error", preparedMessage, Main.instance.Window.Handle);
         }
     }
 }
