--- src/Terraria/Terraria/UI/ItemSlot.cs
+++ src/TerraAngel/Terraria/UI/ItemSlot.cs
@@ -459,7 +_,7 @@
 
         if (Main.keyState.IsKeyDown(Main.FavoriteKey))
         {
-            if (Main.drawingPlayerChat && canShareAt[context])
+            if ((TerraAngel.ClientLoader.ChatWindow?.IsChatting ?? false) && canShareAt[context])
             {
                 Main.cursorOverride = 2;
                 return;
