--- src/Terraria/Terraria/DataStructures/TileReachCheckSettings.cs
+++ src/TerraAngel/Terraria/DataStructures/TileReachCheckSettings.cs
@@ -36,6 +_,12 @@
         {
             y = OverrideYReach.Value;
         }
+
+        if (ToolManager.GetTool<InfiniteReachTool>().Enabled)
+        {
+            x = Main.screenWidth / 32 + 8;
+            y = Main.screenHeight / 32 + 8;
+        }
     }
 
     public void GetTileRegion(Player player, out int LX, out int LY, out int HX, out int HY, int TB = 0)
